File: 937_reorderLogFiles.cpp
//stable: bubble & merge
//unstable: heap & quicksort
// ps c的if后要有括号 然后class的}后要有分号
//referenced from @lzl1234631x
using namespace std;
//Sol1 20ms
bool myCompare(string a, string b){
    int i = a.find(' ');
    int j = b.find(' ');
    if (!isdigit(a[i+1]) && !isdigit(b[j+1]))
        if (a.substr(i+1) == b.substr(j+1))
            return a.substr(0,i+1) < b.substr(0,j+1);
        else
            return a.substr(i+1) < b.substr(j+1);
    else{
        if (!isdigit(a[i+1])) return true;
        return false;
    }
}
class Solution{
public:
    vector<string> reorderLogFiles(vector<string>& logs){
        stable_sort(logs.begin(), logs.end(), myCompare);
        return logs;
    };
};

//Sol2 12ms 确实比python的44ms快多了吖

class Solution{
public:
    vector<string> reorderLogFiles(vector<string>& logs){
        vector<string> digitLogs, ans;
        vector<pair<string, string>> letterLogs;
        for (string &s: logs){
            int i=0;
            while (s[i]!=' ') ++i;
            if (isalpha(s[i+1])) letterLogs.emplace_back(s.substr(0,i), s.substr(i+1));
            //这样不对if (isalpha(s[i+1])) letterLogs.push_back(s.substr(i+1) + " "+s.substr(0,i));
            else digitLogs.push_back(s);
        }
        sort(letterLogs.begin(), letterLogs.end(), [&](auto& a, auto& b){
            return a.second == b.second ? a.first < b.first: a.second < b.second;
        });
        for (auto &p: letterLogs) ans.push_back(p.first + " " + p.second);
        for (string &s: digitLogs) ans.push_back(s);
        return ans;
    }
};

File: 937_reorderLogFiles.py
class Solution:
    def reorderLogFiles(self, logs: List[str]) -> List[str]:
        letter_logs = []
        digit_logs = []
        for i, log in enumerate(logs):
            if (log.split(" ")[1][0]).isdigit():
                digit_logs.append(i)
            else:
                letter_logs.append(i)
        letter_logs.sort(key = lambda idx: (" ".join(logs[idx].split(" ")[1:]), logs[idx].split(" ")[0]))
        return [logs[i] for i in letter_logs] + [logs[i] for i in digit_logs]

        #Sol 1.1 删除引用 48ms -> 44ms
        letter_logs = []
        digit_logs = []
        for i, log in enumerate(logs):
            if (log.split(" ")[1][0]).isdigit():
                digit_logs.append(log)
            else:
                letter_logs.append(log)
        letter_logs.sort(key = lambda log: (" ".join(log.split(" ")[1:]), log.split(" ")[0]))
        return letter_logs + digit_logs

        #Sol 2filter 44ms
        letter_logs = filter(lambda log: log[log.find(" ") + 1].isalpha(), logs)
        digit_logs = filter(lambda log: log[log.find(" ") + 1].isdigit(), logs)
        return sorted(letter_logs, key = lambda log: (log[log.find(" "):], log[:log.find(" ")])) + list(digit_logs)

        '''
        filter举例：https://book.pythontips.com/en/latest/map_filter.html
        number_list = range(-5, 5)
        less_than_zero = list(filter(lambda x: x < 0, number_list))
        '''